Danki2/Assets/Scripts/Abilities/AbilityBuilder.cs

```diff
@@ -0,0 +1,45 @@
+﻿using Abilities;
+using Assets.Scripts.Abilities.InstantCast;
+using System;
+using System.Collections.Generic;
+
+namespace Assets.Scripts.Abilities
+{
+    public enum AbilityReference
+    {
+        Slash,
+        ShieldBash,
+    }
+
+    public class AbilityBuilder
+    {
+        private readonly Func<AbilityContext, Ability> _buildAction;
+
+        public AbilityBuilder(Func<AbilityContext, Ability> buildAction)
+        {
+            _buildAction = buildAction;
+        }
+
+        public Ability Build(AbilityContext context)
+        {
+            return _buildAction(context);
+        }
+
+        // This is where we tie the AbilityReference enum to abilities.
+        private static Dictionary<AbilityReference, AbilityBuilder> abilityBuilders = new Dictionary<AbilityReference, AbilityBuilder>
+        {
+            { AbilityReference.Slash, new AbilityBuilder(c => new Slash(c)) },
+            { AbilityReference.ShieldBash, new AbilityBuilder(c => new ShieldBash(c)) }
+        };
+
+        public static AbilityBuilder GetAbilityBuilder(AbilityReference reference)
```

Comment:
Also do we really need the idea of the AbilityBuilder? Could this class just be AbilityFactory, with this method called GetAbility? Then it could take the ability reference and the context and just call the function in the map for you and return the ability? It seems like that would be simpler and more usable

Reply:
Just saw this comment. Last commit does something similar, but as static methods on Ability.